### ISSUE TITLE
Trying to find using unexpected type

### ISSUE THREAD

**Issue by bolshakov** [hanami/model#287]
I have a JSON API with primary keys of all entities of Postgres UUID type. Users search a lot by ids, and the problem happens when user occasionally enter random string instead of proper UUID.

```ruby
> BooksRepository.find('invalid')
Hanami::Model::InvalidQueryError: PG::InvalidTextRepresentation: ERROR:  invalid input syntax for uuid: "invalid"
LINE 1: SELECT * FROM "books" WHERE ("id" = 'invalid...
```

I'd rather expect this request returns just `nil`, instead of type error. I think controlling such type of things is pretty verbose and annoying.

**Comment by hlegius** [comment]
Hello!

I think there are two concepts here. One is _not found_ a record by given a valid/expected type; In the other hand, there are cases when your client tries to request some data from the API with an `invalid type`.

If I request a record like:

`BooksRepository.find(1234)` and my _expected type_ is an Integer, I should respond `nil` if there is non-existing records for `1234` id. 

if I try to fetch some record with an _invalid or unexpected type_, I should respond with an Exception/Error, because we aren't expecting such type.

From a JSON API point of view [a server may respond with other HTTP status code](http://jsonapi.org/format/#crud-updating-responses-other) like HTTP 422 ([Unprocessable Entity](http://www.restpatterns.org/HTTP_Status_Codes/422_-_Unprocessable_Entity)) which is a good response in cases when your client sends you invalid or malformed data (i.e. a type mismatch).

Postgres throws an Exception which is re-raised by Sequel because you've broken its contract. PG's waiting some __UUID type__ but you sent something else instead.

IMO, if Hanami suppress this kind of exception, it could be impossible to notify/do something to the client that he/she broken the contract.

In my case in particular, I am running a JSON API with Hanami Model just like you and I ended with this solution:

```ruby
    rescue_from Hanami::Model::InvalidQueryError do
      Rack::Response.new(
        [{errors: [{source: 'Entity does not support format received'}]}.to_json],
        422,
        {'Content-type' => 'application/vnd.api+json'}
      ).finish
    end
``` 

In my case I am using Grape to deliver content.

/cc @hanami/core-team

**Comment by bolshakov** [comment]
I understand that, it is may be counterintuitive to suppress such errors.  That how behaves ActiveRecord in the same case:

```ruby
Book.find('unexpected type')
ActiveRecord::RecordNotFound: Couldn't find Book with 'id'=unexpected type
```

But it performs convention string to UUID and returns nil value. Here is the generated SQL:

```
SELECT  "books".* FROM "books" WHERE "books"."id" = $1 LIMIT 1  [["id", nil]]
```

Hanami in some cases performs conversation too. For example if I have integer primary keys, it is possible to successfully  perform both queries:

* `BooksRepository.find(1)` 
* `BooksRepository.find('1')`

**Comment by jodosha** [comment]
Let's start from the beginning. What PG does if I try to run that query?

```sql
bookshelf_development=# select * from books;
                  id                  | title
--------------------------------------+-------
 970b24b7-f6f7-4958-872a-2dbf970ae4a1 | TDD
(1 row)

bookshelf_development=# select * from books where id='invalid';
ERROR:  invalid input syntax for uuid: "invalid"
LINE 1: select * from books where id='invalid';
                                     ^
```

This is the reason why Model raises a `Hanami::Model::InvalidQueryError: PG::InvalidTextRepresentation` exception.

---

That's true that we accept both string and integer representations for ID, when they are stored as an integer in the database too.

Now, this isn't Model, but Ruby.

We send SQL queries as strings to the database. When we interpolate a string with an integer, the output is a string.

```ruby
irb(main):001:0> id = 1
=> 1
irb(main):002:0> %(SELECT * FROM books WHERE id='#{ id }')
=> "SELECT * FROM books WHERE id='1'"
```

This isn't accurate, but it's definitely useful because we don't have to coerce the correct type to run a query. For instance we leave id as a string, if it comes from an HTTP request, without the need to cast to an int.

To recap:

```ruby
BookRepository.find(1)   # => book
BookRepository.find('1') # => book

BookRepository.find(999) # => nil
BookRepository.find('999') # => nil

BookRepository.find('invalid')
# => TypeError: can't convert "invalid" into Integer
from /Users/luca/.gem/ruby/2.3.0/gems/hanami-utils-0.7.0/lib/hanami/utils/kernel.rb:325:in `rescue in Integer'
```

---

I think that this last error shouldn't be there. The semantic of `.find` is: "please find a record with this primary key and return nil in case it's missing".

That is true for the first two examples, but this contract is broken when we use `"invalid"` as ID.

---

We should make it to work, regardless of the underlying database/Ruby representation of the ID. Thoughts?

**Comment by jodosha** [comment]
I forgot to mention that the example reported by @bolshakov is useful for testing purposes. If I want to assert that my endpoint returns a `404` for an unknown ID, I'd test with `GET /books/unknown` and it's _"natural"_ to expect a `404`, not a `500`.

Getting back to the lower level where Model belongs. To make this example to work, a repository should return `nil`, instead of raise an error.`

**Comment by Bounga** [comment]
> That is true for the first two examples, but this contract is broken when we use "invalid" as ID.
> We should make it to work, regardless of the underlying database/Ruby representation of the ID. Thoughts?

@jodosha :+1: I agree it should be consistent.

> If I want to assert that my endpoint returns a 404 for an unknown ID, I'd test with GET /books/unknown and it's "natural" to expect a 404, not a 500.

That's what I would do too, you shouldn't get a `500` because ID is "invalid". Repository should handle this for us since it's a pretty common case.

**Comment by hlegius** [comment]
@jodosha I am working with these specific errors (see #284). If we will keep the consistence by returning `nil` here, I can deal with this one as well. o/
